sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -90,25 +95,101 @@ trait BaseLimitExec extends UnaryExecNode with CodegenSupport {
+      val submittedStageFuture = sparkContext.submitMapStage(shuffleDependency)
+      submittedStageFuture.get().numberOfOutput.toSeq
+    } else {
+      Nil
+    }
 
-  override def requiredChildDistribution: List[Distribution] = AllTuples :: Nil
+    // Try to keep child plan's original data parallelism or not. It is enabled by default.
+    val respectChildParallelism = sqlContext.conf.enableParallelGlobalLimit
 
-  override def outputPartitioning: Partitioning = child.outputPartitioning
+    val shuffled = new ShuffledRowRDD(shuffleDependency)
 
-  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+    val sumOfOutput = numberOfOutput.sum
+    if (sumOfOutput <= limit) {
+      shuffled
+    } else if (!respectChildParallelism) {
+      // This is mainly for tests.
+      // We take the rows of each partition until we reach the required limit number.
+      var countForRows = 0
+      val takeAmounts = new mutable.HashMap[Int, Int]()
+      numberOfOutput.zipWithIndex.foreach { case (num, index) =>
+        if (countForRows + num < limit) {
+          countForRows += num
+          takeAmounts += ((index, num))
+        } else {
+          val toTake = limit - countForRows
+          countForRows += toTake
+          takeAmounts += ((index, toTake))
+        }
+      }
+      val broadMap = sparkContext.broadcast(takeAmounts)
+      shuffled.mapPartitionsWithIndexInternal { case (index, iter) =>
+        broadMap.value.get(index).map { size =>
+          iter.take(size)
+        }.get
+      }
+    } else {
+      // We try to distribute the required limit number of rows across all child rdd's partitions.
+      var numToReduce = (sumOfOutput - limit)
+      val reduceAmounts = new mutable.HashMap[Int, Int]()
+      val nonEmptyParts = numberOfOutput.filter(_ > 0).size
+      val reducePerPart = numToReduce / nonEmptyParts
+      numberOfOutput.zipWithIndex.foreach { case (num, index) =>
+        if (num >= reducePerPart) {
+          numToReduce -= reducePerPart
+          reduceAmounts += ((index, reducePerPart))
+        } else {
+          numToReduce -= num
+          reduceAmounts += ((index, num))
+        }
+      }
+      while (numToReduce > 0) {
+        numberOfOutput.zipWithIndex.foreach { case (num, index) =>
+          val toReduce = if (numToReduce / nonEmptyParts > 0) {
+            numToReduce / nonEmptyParts
+          } else {
+            numToReduce
+          }
+          if (num - reduceAmounts(index) >= toReduce) {
+            reduceAmounts(index) = reduceAmounts(index) + toReduce
+            numToReduce -= toReduce
+          } else if (num - reduceAmounts(index) > 0) {
+            reduceAmounts(index) = reduceAmounts(index) + 1
+            numToReduce -= 1
+          }
+        }
+      }
+      val broadMap = sparkContext.broadcast(reduceAmounts)
+      shuffled.mapPartitionsWithIndexInternal { case (index, iter) =>
+        broadMap.value.get(index).map { size =>
+          iter.drop(size)
```

Comment:
Maybe we should replace current logic to calculate the number of rows to take for each partition.